### PR TITLE
Bias Mitigation and Direction Methods

### PR DESCRIPTION
**Pull request by ArjunSubramonian**
Additions proposed in this pull request:

- Added four bias direction methods (`PCABiasDirection`, `PairedPCABiasDirection`, `TwoMeansBiasDirection`, `ClassificationNormalBiasDirection`) and four bias mitigation methods (`LinearBiasMitigator`, `HardBiasMitigator`, `INLPBiasMitigator`, `OSCaRBiasMitigator`)

### CODE REVIEW

File: allennlp/fairness/bias_mitigators.py [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+            mask = ~(evaluation_embeddings == 0).all(dim=-1)
+            # Transform all evaluation embeddings
+            # using orthonormal basis computed above
+            rotated_evaluation_embeddings = torch.matmul(
+                evaluation_embeddings[mask], rotation_matrix.t()
+            )
+            # Want to adjust first 2 coordinates and leave d - 2
+            # other orthogonal components fixed
+            fixed_rotated_evaluation_embeddings = rotated_evaluation_embeddings[..., 2:]
+            # Restrict attention to subspace S
```
Comment (AkshitaB) [inline allennlp/fairness/bias_mitigators.py:449]: where subspace S is ...?

File: allennlp/fairness/bias_direction.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+            A unit tensor of size (dim, ) representing the concept subspace.
+        """
+
+        # Some sanity checks
+        if seed_embeddings.ndim < 2:
+            raise ConfigurationError("seed_embeddings1 must have at least two dimensions.")
+
+        with torch.set_grad_enabled(self.requires_grad):
+            # pca_lowrank centers the embeddings by default
+            _, _, V = torch.pca_lowrank(seed_embeddings, q=2)
```
[the remark this answers is not in the document]
Comment (ArjunSubramonian) [reply allennlp/fairness/bias_direction.py:72]: I followed the VERB implementation + paper. I think the intuition behind this is that there will be two dimensions when applying PCA to definitionally-gendered words: 1) the gender direction, 2) all other directions, with the gender direction being principal.

File: allennlp/fairness/bias_mitigators.py [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+            characterize gender in real life.
+
+        # Parameters
+
+        evaluation_embeddings : `torch.Tensor`
+            A tensor of size (evaluation_batch_size, ..., dim) of embeddings for which to mitigate bias.
+        bias_direction : `torch.Tensor`
+            A unit tensor of size (dim, ) representing the concept subspace. The words
+            that are used to define the bias direction are considered definitionally
+            gendered and not modified.
```
[the remark this answers is not in the document]
Comment (ArjunSubramonian) [reply allennlp/fairness/bias_mitigators.py:84]: Yes!

File: allennlp/fairness/bias_mitigators.py [hunk 4]
```diff
@@ -0,0 +1,542 @@
...
+
+    def _remove_component(
+        self, embeddings: torch.Tensor, bias_direction: torch.Tensor, normalize: bool = False
+    ):
+        return embeddings - self._proj(embeddings, bias_direction, normalize)
+
+
+class HardBiasMitigator(BiasMitigator):
+    """
+    Hard bias mitigator. Mitigates bias in embeddings by:
```
[the remark this answers is not in the document]
Comment (ArjunSubramonian) [reply allennlp/fairness/bias_mitigators.py:40]: Added note at top of both mitigator and direction files.

File: allennlp/fairness/bias_mitigators.py [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+
+    1. Neutralizing: ensuring protected variable-neutral words remain equidistant
+    from the bias direction by removing component of embeddings
+    in the bias direction.
+
+    2. Equalizing: ensuring that protected variable-related words are averaged
+    out to have the same norm.
+
```
[the remark this answers is not in the document]
Comment (ArjunSubramonian) [reply allennlp/fairness/bias_mitigators.py:48]: For each mitigation method, I just linked the appropriate figure in the VERB paper, as I think the pictures are the most helpful.

File: allennlp/fairness/bias_direction.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+        !!! Note
+            seed_embeddings1 and seed_embeddings2 need NOT be the same size. Furthermore,
+            the embeddings at the same positions in each of seed_embeddings1 and seed_embeddings2
+            are NOT expected to form seed word pairs.
+
+        !!! Note
+            All tensors are expected to be on the same device.
+
+        !!! Note
+            This bias direction method is NOT differentiable.
```
[the remark this answers is not in the document]
Comment (ArjunSubramonian) [reply allennlp/fairness/bias_direction.py:272]: Yes, this is part of the bias mitigators and direction wrappers PR - this PR is just the functional API.

File: tests/fairness/bias_mitigators_test.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+            evaluation_embeddings.append(torch.Tensor(emb_data[word]).reshape(1, -1))
+            expected_bias_mitigated_embeddings.append(
+                torch.Tensor(emb_data["linear_two_means_" + word]).reshape(1, -1)
+            )
+        self.evaluation_embeddings = torch.cat(evaluation_embeddings).reshape(2, 2, -1)
+        self.expected_bias_mitigated_embeddings = torch.cat(
+            expected_bias_mitigated_embeddings
+        ).reshape(2, 2, -1)
+
+    def teardown_method(self):
```
[the remark this answers is not in the document]
Comment (ArjunSubramonian) [reply tests/fairness/bias_mitigators_test.py:51]: we shouldn't :) I just forgot to call the parent setup_method(), so the tmp dir wasn't being deleted.

File: allennlp/fairness/bias_direction.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ArjunSubramonian) [reply allennlp/fairness/bias_direction.py:72]: Added a comment in the file itself

File: allennlp/fairness/bias_mitigators.py (hunk 1 above)
Comment (ArjunSubramonian) [reply allennlp/fairness/bias_mitigators.py:449]: the subspace spanned by the bias directions (made a comment in the file)

**Pull request merged**